File: android-client/lib/services/replay_service.dart
import 'dart:async';

import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_database/firebase_database.dart';
import 'package:flutter/material.dart';
import 'package:rxdart/rxdart.dart';

import '../classes/vec2.dart';

class ReplayEvent {
  String type;
  double timestamp;
  int player;
  Map<String, dynamic> data;

  ReplayEvent(this.type, this.timestamp, this.player, this.data);

  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'timestamp': timestamp,
      'player': player,
      'data': data,
    };
  }

  factory ReplayEvent.fromJson(Map<dynamic, dynamic> json) {
    return ReplayEvent(
      json['type'] as String,
      (json['timestamp'] as num).toDouble(),
      json['player'] as int,
      Map<String, dynamic>.from(json['data']),
    );
  }
}

class ReplayService {
  final BehaviorSubject<dynamic> blinkEmit = BehaviorSubject<dynamic>();
  final BehaviorSubject<dynamic> resetImage = BehaviorSubject<dynamic>();
  final BehaviorSubject<dynamic> drawError = BehaviorSubject<dynamic>();
  final BehaviorSubject<dynamic> timePositionChanged =
      BehaviorSubject<dynamic>();

  int progressIndex = 0;
  double progressTime = 0;
  double gameTime = 0;
  List<ReplayEvent> events = [];
  bool isDisplaying = false;
  Timer? timerInterval;
  double replaySpeedValue = 1;
  bool isPause = false;
  double startGameTime = 0;
  int interval = 10;
  double timeBuffer = 1;
  bool isInReplayMode = false;
  bool isSavedReplay = false;

  late BuildContext context;

  late String uid;
  late String mapId;

  int getGameTime() {
    return gameTime.toInt().floor();
  }

  int getProgressTime() {
    return progressTime.toInt().floor();
  }

  double getTimeStamp() {
    return (DateTime.now().millisecondsSinceEpoch.toDouble() - startGameTime) /
        1000;
  }

  void logChat(int player, String message) {
    print("log Chat: $player, $message");
    events.add(ReplayEvent('chat', getTimeStamp(), player, {
      'value': message,
      'time': DateTime.now().millisecondsSinceEpoch.toDouble()
    }));
  }

  void logGlobalMessage(String message) {
    print("log Global Message: $message");
    events.add(
        ReplayEvent('globalMessage', getTimeStamp(), -1, {'value': message}));
  }

  void logClick(int player, int diff) {
    print("log Click: $player, $diff");
    events.add(ReplayEvent('click', getTimeStamp(), player, {
      'diff': diff,
      'time': DateTime.now().millisecondsSinceEpoch.toDouble()
    }));
  }

  void logError(int player, Vec2 coords) {
    print("log Error: $player, $coords");
    events.add(ReplayEvent('error', getTimeStamp(), player, {
      'coords': {'res': -1, 'x': coords.x, 'y': coords.y},
      'time': DateTime.now().millisecondsSinceEpoch.toDouble()
    }));
  }

  void logOtherError(int player) {
    print("log Other Error: $player");
    events.add(ReplayEvent('otherError', getTimeStamp(), player,
        {'time': DateTime.now().millisecondsSinceEpoch.toDouble()}));
  }

  void replayOrResume() {
    isInReplayMode = true;
    print(events);
    timerInterval = Timer.periodic(Duration(milliseconds: interval), (Timer t) {
      print(
          "XX 1 - ${progressTime >= gameTime + timeBuffer}, $progressTime, $gameTime, $timeBuffer");
      print(
          "XX 2 - ${progressIndex >= events.length}, $progressIndex, ${events.length}");

      progressTime += (interval * replaySpeedValue) / 1000;
      timePositionChanged.add(null);
      if (progressIndex < events.length) {
        ReplayEvent event = events[progressIndex];
        print(
            "progressTime: $progressTime, eventTime: ${event.timestamp}, event");
        print('event type is: ${event.type} $progressIndex');
        if (progressTime >= event.timestamp - timeBuffer) {
          switch (event.type) {
            case 'chat':
              print('playing chat');
              blinkEmit.add(event);
              break;
            case 'click':
              print('playing click');
              blinkEmit
                  .add({'diff': event.data['diff'], 'speed': replaySpeedValue});
              break;
            case 'error':
              print('playing error');
              drawError.add({
                'coords': event.data['coords'],
                'player': event.player,
                'time': event.data['time']
              });
              break;
            case 'playing otherError':
              print('other error');
              break;
            case 'playing globalMessage':
              print('globalMessage');
              blinkEmit.add(event);
              break;
          }
          progressIndex++;
        }
      }

      if (progressTime >= gameTime + timeBuffer) {
        print("XX - replay over");
        openEndDialog();
        pause();
      }
    });
  }

  void seekToTime(double newTime) {
    pause();
    progressTime = newTime;
    progressIndex =
        events.indexWhere((event) => event.timestamp >= newTime / 1000);
    if (progressIndex == -1) {
      progressIndex = 0;
    }
    events =
        events.where((event) => event.timestamp >= newTime / 1000).toList();
    if (!isPause) {
      resume();
    }
  }

  void setBuildContext(BuildContext context) {
    this.context = context;
  }

  void openEndDialog() {
    // this.customDialogService
    // .openDialog({
    //     title: 'Reprise de la partie terminée',
    //     confirm: 'Recommencer la reprise vidéo',
    //     cancel: 'Revenir à la page principale',
    // })
    // .afterClosed()
    // .subscribe((result: boolean) => {
    //     if (result) {
    //         this.reset();
    //         this.resume();
    //     } else {
    //         this.zone.run(() => {
    //             this.isDisplaying = false;
    //             this.reset();
    //             this.router.navigate(['/classic']);
    //         });
    //     }
    // });

    showDialog(
      context: context,
      builder: (BuildContext dialogContext) {
        return AlertDialog(
          title: Text('Reprise de la partie terminée'),
          actions: [
            TextButton(
              onPressed: () {
                reset();
                Navigator.of(dialogContext).pop();
                Navigator.of(context).pop();
              },
              child: Text('Revenir à la page principale'),
            ),
            TextButton(
              onPressed: () {
                Navigator.of(dialogContext).pop();
                reset();
                resume();
              },
              child: Text('Recommencer la reprise vidéo'),
            ),
            if (!isSavedReplay)
              TextButton(
                onPressed: () {
                  saveReplay();
                  Navigator.of(dialogContext).pop();
                  Navigator.of(context).pop();
                },
                child: Text('Sauvegarder la reprise'),
              ),
          ],
        );
      },
    );
  }

  Future<void> saveReplay() async {
    await Firebase.initializeApp();
    final db = FirebaseDatabase.instance.reference();
    final userRef = db.child('users').child(uid).child('replayEvents');

    final replayRef = userRef.push();

    final eventsJson = events.map((e) => e.toJson()).toList();

    replayRef.set({
      'startGameTime': startGameTime,
      'gameTime': gameTime,
      'events': eventsJson,
      'mapId': mapId
    }).then((_) {
      print('Replay events saved successfully!');
    }).catchError((error) {
      print('Error saving replay events: $error');
    });
  }

  void replay() {
    isInReplayMode = true;
    replayOrResume();
  }

  void pause() {
    isPause = true;
    timerInterval?.cancel();
  }

  void resume() {
    isPause = false;
    replayOrResume();
  }

  void toggleReplay() {
    isPause ? resume() : pause();
  }

  void setGameTime(double time) {
    gameTime = time / 1000;
  }

  void saveStartTime() {
    startGameTime = DateTime.now().millisecondsSinceEpoch.toDouble();
  }

  void reset() {
    progressIndex = 0;
    progressTime = 0;
    resetImage.add(null);
  }

  void dispose() {
    progressIndex = 0;
    progressTime = 0;
    gameTime = 0;
    events = [];
    timerInterval?.cancel();
    timerInterval = null;
    replaySpeedValue = 1;
    isPause = false;
    startGameTime = 0;
    isDisplaying = false;
    isInReplayMode = false;
  }
}

File: desktop-client/src/app/services/replay.service.ts
import { EventEmitter, Injectable, NgZone, OnDestroy, OnInit } from '@angular/core';
import { Router } from '@angular/router';
import { PlayerIndex } from '@app/components/game-data';
import { ReplayEvent } from '@app/interfaces/replay-event';
import { MessageType } from '@app/services/game-message';
import { LocalMessagesService } from '@app/services/local-messages.service';
import { getDatabase, push, ref, set } from 'firebase/database';
import { AuthService } from './auth-service';
import { ClassicGameLogicService } from './classic-game-logic.service';
import { CustomDialogService } from './custom-dialog.service';
import { SocketClientService } from './socket-client.service';

@Injectable({
    providedIn: 'root',
})
export class ReplayService implements OnDestroy, OnInit {
    blinkEmit = new EventEmitter<any>();
    resetImage = new EventEmitter<any>();
    drawError = new EventEmitter<any>();
    progressIndex: number = 0;
    progressTime: number | any = 0;
    gameTime: number;
    events: ReplayEvent[] = [];
    isDisplaying: boolean = false;
    timerInterval: any;
    replaySpeedValue: number = 1;
    isPause: boolean = false;
    startGameTime: number;
    interval = 10;
    timeBuffer: number = 1;
    timePositionChanged = new EventEmitter<number>();
    isSavedReplay = false;

    uid: string | null;

    // eslint-disable-next-line max-params
    constructor(
        public socket: SocketClientService,
        public localMessages: LocalMessagesService,
        public customDialogService: CustomDialogService,
        private zone: NgZone,
        public router: Router,
        public authService: AuthService,
        public gameService: ClassicGameLogicService,
    ) {
        if (!sessionStorage.getItem('user_uid')) {
            this.uid = this.authService.getUID();
            sessionStorage.setItem('user_uid', this.uid);
            console.log('hihihiih', this.uid);
        } else {
            this.uid = sessionStorage.getItem('user_uid');
            console.log('hihihiih2', this.uid);
        }
    }

    async ngOnInit(): Promise<void> {
        if (!sessionStorage.getItem('user_uid')) {
            this.uid = this.authService.getUID();
            sessionStorage.setItem('user_uid', this.uid);
            console.log('hihihiih', this.uid);
        } else {
            this.uid = sessionStorage.getItem('user_uid');
            console.log('hihihiih2', this.uid);
        }
    }

    getGameTime() {
        return Math.floor(this.gameTime);
    }
    getProgressTime() {
        return Math.floor(this.progressTime);
    }

    logChat(player: PlayerIndex, value: string) {
        this.events.push({
            type: 'chat',
            timestamp: (new Date().getTime() - this.startGameTime) / 1000,
            player,
            data: { value, time: new Date().toTimeString().slice(0, 8) },
        });
    }

    logGlobalMessage(value: string) {
        this.events.push({
            type: 'globalMessage',
            timestamp: (new Date().getTime() - this.startGameTime) / 1000,
            player: PlayerIndex.None,
            data: { value },
        });
    }

    logClick(player: PlayerIndex, diff: any) {
        this.events.push({
            type: 'click',
            timestamp: (new Date().getTime() - this.startGameTime) / 1000,
            player,
            data: { diff, time: new Date().toTimeString().slice(0, 8) },
        });
    }

    logError(player: PlayerIndex, coords: any) {
        this.events.push({
            type: 'error',
            timestamp: (new Date().getTime() - this.startGameTime) / 1000,
            player,
            data: { coords, time: new Date().toTimeString().slice(0, 8) },
        });
    }

    logOtherError(player: PlayerIndex) {
        this.events.push({
            type: 'otherError',
            timestamp: (new Date().getTime() - this.startGameTime) / 1000,
            player,
            data: { time: new Date().toTimeString().slice(0, 8) },
        });
    }

    replayOrResume() {
        console.log('events', this.events);
        this.timerInterval = setInterval(() => {
            this.progressTime += (this.interval * this.replaySpeedValue) / 1000;
            const event: ReplayEvent = this.events[this.progressIndex];
            if (event) {
                if (this.progressTime >= event.timestamp - this.timeBuffer) {
                    switch (event.type) {
                        case 'chat':
                            this.localMessages.addChatMessage(event.player, event.data.value);
                            break;
                        case 'globalMessage':
                            this.localMessages.add(event.data.value, event.data.time);
                            break;
                        case 'click':
                            this.blinkEmit.emit({ diff: event.data.diff, speed: this.replaySpeedValue });
                            this.localMessages.addMessage(MessageType.DifferenceFound, event.player, event.data.time);
                            break;
                        case 'error':
                            this.localMessages.addMessage(MessageType.Error, event.player, event.data.time);
                            this.drawError.emit(event.data.coords);
                            break;
                        case 'otherError':
                            this.localMessages.addMessage(MessageType.Error, event.player, event.data.time);
                            break;
                    }
                    this.progressIndex++;
                }
            }
            if (this.progressTime >= this.gameTime + this.timeBuffer) {
                this.openEndDialog();
                this.pause();
            }
        }, this.interval);
    }

    // -----

    // updateProgressTime(newTime: number) {
    //     this.pause();
    //     this.progressTime = newTime;
    //     this.progressIndex = this.events.findIndex((event) => event.timestamp >= newTime / 1000);
    //     if (!this.isPause) {
    //         this.resume();
    //     }
    //     this.timePositionChanged.emit(this.progressTime);
    // }

    seekToTime(newTime: number) {
        this.pause();
        this.progressTime = newTime;
        this.progressIndex = this.events.findIndex((event) => event.timestamp >= newTime / 1000);
        if (this.progressIndex === -1) {
            this.progressIndex = 0;
        }
        this.events = this.events.filter((event) => event.timestamp >= newTime / 1000);
        if (!this.isPause) {
            this.resume();
        }
    }

    // -----

    openEndDialog() {
        this.customDialogService
            .openDialog({
                title: 'Reprise de la partie terminée',
                confirm: 'Recommencer la reprise vidéo',
                cancel: 'Revenir à la page principale',
                saveReplay: this.isSavedReplay ? undefined : 'Sauvegarder la reprise',
            })
            .afterClosed()
            .subscribe((result: boolean) => {
                if (result === true) {
                    this.reset();
                    this.resume();
                } else if (result === false) {
                    this.zone.run(() => {
                        this.isDisplaying = false;
                        this.reset();
                        this.router.navigate(['/classic']);
                    });
                } else if (result === 'saveReplay') {
                    console.log('saving', this.uid, this.gameService.map?.id);
                    this.saveReplay(this.uid ?? '', this.gameService.map?.id ?? '');
                    this.zone.run(() => {
                        this.router.navigate(['/classic']);
                    });
                }
            });
    }

    replay() {
        this.replayOrResume();
    }

    pause() {
        this.isPause = true;
        clearInterval(this.timerInterval);
    }

    resume() {
        this.isPause = false;
        this.replayOrResume();
    }

    toggleReplay() {
        if (this.isPause) this.resume();
        else this.pause();
    }

    setGameTime(time: number) {
        this.gameTime = time / 1000;
    }

    saveStartTime() {
        this.startGameTime = new Date().getTime();
    }

    reset() {
        this.progressIndex = 0;
        this.progressTime = 0;
        this.localMessages.reset();
        this.resetImage.emit();
    }

    saveReplay(uid: string, mapId: string) {
        const db = getDatabase();
        const userRef = ref(db, `users/${uid}/replayEvents`);

        const replayRef = push(userRef);

        set(replayRef, { startGameTime: this.startGameTime, gameTime: this.gameTime, events: this.events, mapId: mapId })
            .then(() => {
                console.log('Replay events saved successfully!');
            })
            .catch((error: any) => {
                console.error('Error saving replay events:', error);
            });
    }

    ngOnDestroy() {
        this.progressIndex = 0;
        this.progressTime = 0;
        this.gameTime = 0;
        this.events = [];
        clearInterval(this.timerInterval);
        this.timerInterval = null;
        this.replaySpeedValue = 1;
        this.isPause = false;
        this.startGameTime = 0;
        this.isDisplaying = false;
        this.isSavedReplay = false;
    }
}
